offscreen/scripts/lattice.ts

```diff
@@ -0,0 +1,86 @@
+import {
+  KnownOrigins,
+  OffscreenCommunicationTarget,
+} from '../../shared/constants/offscreen-communication';
+
+async function openConnectorTab(url: string) {
+  const browserTab = window.open(url);
+  if (!browserTab) {
+    throw new Error('Failed to open Lattice connector.');
+  }
+
+  return browserTab;
+}
+
+export default function init() {
+  /**
+   * The main element of the eth-lattice-keyring library that is impacted by MV3
+   * is the _openConnectorTab method which is responsible for opening a tab to
+   * the lattice connector. This method is called by the _getCreds method of the
+   * keyring. In the default keyring this would be attempted inside the service
+   * worker which has no DOM and therefore would fail. The solution is to split
+   * the functionality so that the openConnectorTab portion operates inside the
+   * offscreen document with its DOM. That is what this file is responsible for.
+   *
+   * When receiving a message from the service worker script targeting the
+   * lattice iframe, this listener will execute and open the new tab for
+   * connecting to the lattice device. The response from the
+   */
+  chrome.runtime.onMessage.addListener((msg, _sender, sendResponse) => {
+    if (msg.target !== OffscreenCommunicationTarget.latticeOffscreen) {
+      return;
+    }
+
+    // Open the tab
+    openConnectorTab(msg.params.url).then((browserTab) => {
+      // Watch for the open window closing before creds are sent back
+      const listenInterval = setInterval(() => {
+        if (browserTab.closed) {
+          clearInterval(listenInterval);
+          sendResponse({
+            error: new Error('Lattice connector closed.'),
+          });
+        }
+      }, 500);
+
+      // On a Chromium browser we can just listen for a window message from the
+      // lattice tab (using the event.origin property). When we get that it'll be
+      // the response from the lattice connector with the deviceID and password.
+      // We can then forward that response to the lattice-offscreen-keyring's
+      // _getCreds method using sendResponse API from the chrome runtime.
+      window.addEventListener(
+        'message',
+        (event) => {
+          // Ensure origin
+          if (event.origin !== KnownOrigins.lattice) {
+            return;
+          }
+
+          try {
+            // Stop the listener
+            clearInterval(listenInterval);
+
+            // Parse and return creds
+            const creds = JSON.parse(event.data);
```

Comment:
password? Does that mean any code running in this realm can listen to this message and get their hands on it as well? What does this password represent?

Reply:
I need to investigate further this week and get back to you on this, but I belive this code was lifted to this area from where it existed before in the stack in MV2.

Reply:
I'm sure this PR is mostly just some reorganization of code, but this pattern caught my eye so I just wanna make sure it's safe.

Reply:
This is the original code, from eth-lattice-keyring:

```
  _getCreds() {
    return new Promise((resolve, reject) => {
      // We only need to setup if we don't have a deviceID
      if (this._hasCreds())
        return resolve();
      // If we are not aware of what Lattice we should be talking to,
      // we need to open a window that lets the user go through the
      // pairing or connection process.
      const name = this.appName ? this.appName : 'Unknown'
      const base = 'https://lattice.gridplus.io';
      const url = `${base}?keyring=${name}&forceLogin=true`;
      let listenInterval;

      // PostMessage handler
      function receiveMessage(event) {
        // Ensure origin
        if (event.origin !== base)
          return;
        try {
          // Stop the listener
          clearInterval(listenInterval);
          // Parse and return creds
          const creds = JSON.parse(event.data);
          if (!creds.deviceID || !creds.password)
            return reject(new Error('Invalid credentials returned from Lattice.'));
          return resolve(creds);
        } catch (err) {
          return reject(err);
        }
      }

      // Open the tab
      this._openConnectorTab(url)
      .then((conn) => {
        if (conn.chromium) {
          // On a Chromium browser we can just listen for a window message
          window.addEventListener("message", receiveMessage, false);
          // Watch for the open window closing before creds are sent back
          listenInterval = setInterval(() => {
            if (conn.chromium.closed) {
              clearInterval(listenInterval);
              return reject(new Error('Lattice connector closed.'));
            }
          }, 500);
```

This is the part that was spliced out into our custom LatticeKeyring existing in lattice-offscreen-keyring.ts, which extends the LatticeKeyring class and overwrites the _getCreds() method to send the request to the offscreen document so it can have dom access. 

Nothing has changed in how we connect to the lattice, just that it now happens in the offscreen document instead. The window and communication via event listeners is present in mv2 today.  I would probably compare the _getCreds here with https://github.com/GridPlus/eth-lattice-keyring/blob/master/index.js#L461 to see if any changes I made increase our risk.

Reply:
Makes sense. I guess under a lockdown env this shouldn't be a problem. Thanks you for the explanation.